src/paths/random_walk.c

```diff
@@ -301,3 +240,152 @@ igraph_error_t igraph_random_edge_walk(const igraph_t *graph,
+ *   that at least one strictly positive weight is found among the
+ *   outgoing edges of each vertex. Additionally, no edge weight may
+ *   be NaN. If either case does not hold, an error is returned. If it
+ *   is a NULL pointer, all edges are considered to have equal weight.
+ * \param vertices An allocated vector, the result is stored here as
+ *   a list of vertex IDs. It will be resized as needed.
+ *   It includes vertex ids of starting and ending vertices.
+ *   Length of the vertices vector: \p steps + 1
+ * \param edges An initialized vector, the indices of traversed
+ *   edges are stored here. It will be resized as needed.
+ *   Length of the edges vector: \p steps
+ * \param start The start vertex for the walk.
+ * \param steps The number of steps to take. If the random walk gets
+ *   stuck, then the \p stuck argument specifies what happens.
+ *   \p steps is the number of edges to traverse on the walk.
+ * \param mode How to walk along the edges in directed graphs.
+ *   \c IGRAPH_OUT means following edge directions, \c IGRAPH_IN means
+ *   going opposite the edge directions, \c IGRAPH_ALL means ignoring
+ *   edge directions. This argument is ignored for undirected graphs.
+ * \param stuck What to do if the random walk gets stuck.
+ *   \c IGRAPH_RANDOM_WALK_STUCK_RETURN means that the function returns
+ *   with a shorter walk (for \p vertices and \p edges);
+ *   \c IGRAPH_RANDOM_WALK_STUCK_ERROR means that an error is reported;
+ *   In both cases \p vertices and \p edges are truncated to contain
+ *   the actual interrupted walk.
+ * \return Error code.
+ *
+ * Time complexity:
+ *   For unweighted graph: O(l + d),
+ *   For weighted graph: O(l * log(k) + d),
+ *   where \c l is the length of the walk, \c d is the total degree of the visited nodes
+ *   and \c k is the average degree of vertices of the given graph.
+ */
+
+
+igraph_error_t igraph_random_walk(const igraph_t *graph,
+                       const igraph_vector_t *weights,
+                       igraph_vector_int_t *vertices,
+                       igraph_vector_int_t *edges,
+                       igraph_integer_t start,
+                       igraph_neimode_t mode,
+                       igraph_integer_t steps,
+                       igraph_random_walk_stuck_t stuck) {
+
+    /* TODO:
+       - multiple walks potentially from multiple start vertices
+    */
+
+    igraph_integer_t vc = igraph_vcount(graph);
+    igraph_integer_t ec = igraph_ecount(graph);
+    
+    if (!(mode == IGRAPH_ALL || mode == IGRAPH_IN || mode == IGRAPH_OUT)) {
+        IGRAPH_ERROR("Invalid mode parameter.", IGRAPH_EINVMODE);
+    }
+    
+    if (start < 0 || start >= vc) {
+        IGRAPH_ERRORF("Starting vertex must be between 0 and the "
+                      "number of vertices in the graph (%" IGRAPH_PRId
+                      "), got %" IGRAPH_PRId ".", IGRAPH_EINVAL,
+                      vc, start);
+    }
+    if (steps < 0) {
```

Comment:
I'm thinking briefly about the edge case for which `steps == 0`. In that case, `vertices[0]` will not be valid, but this is now not caught. Probably the easiest is just to add an `if (steps == 0)` on l. 336 just before calling the relevant subroutines to return without doing anything more.

Reply:
Sure. I've added it. Also, added
```
* If \p steps is 0, the function returns without doing anything more.
```
under the description of steps params in doc.

Reply:
Thanks! Not sure if that's needed though, it's also implicit that if 0 steps are being taken, that you will simply get vectors of length 0 and 1 respectively right, so in a sense there's nothing special about `steps == 0`, we just need to make sure it works correctly.